### PR TITLE
Add konko llm integration

### PR DESCRIPTION
**Pull request by mahaddad**
# Description

_Thank you to the LlamaIndex team for the great project and in advance for your review. Let me know if I can provide any other additional information or do things differently in the future to make your lives easier 🙏 _

The [Konko API](konko.ai) is a fully managed API designed to help application developers:

- Select the right LLM(s) for their application
- Prototype with various open-source and proprietary LLMs
- Move to production in-line with their security, privacy, throughput, latency SLAs without infrastructure set-up or administration using Konko AI's SOC 2 compliant infrastructure

We attempted to minimize code duplication where possible, but in some instances needed to recreate certain functions because we required the ability to handle multiple api keys. 

- [✔️  ] New feature (non-breaking change which adds functionality)
- [ ✔️ ] This change requires a documentation update (we've included documentation)

# How Has This Been Tested?

To run our test files, please `pip install konko`, otherwise the tests will be skipped.

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ✔️ ] Added new unit/integration tests
- [ ✔️ ] Added new notebook (that tests end-to-end)
- [✔️  ] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [✔️  ] I have performed a self-review of my own code
- [ ✔️ ] I have commented my code, particularly in hard-to-understand areas
- [ ✔️ ] I have made corresponding changes to the documentation
- [ ✔️ ] My changes generate no new warnings
- [ ✔️ ] I have added tests that prove my fix is effective or that my feature works
- [ ✔️ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: llama_index/llms/konko.py
Comment (logan-markewich) [inline llama_index/llms/konko.py:1]: High level comment -- this seems to be a lot of code duplication. Is it possible to use the existing openai class and just changing the base API URL?

**Comment by mahaddad** [comment]
We tried to use the OpenAI class initially but faced issues supporting
multiple API keys. We'll take another look to double check because I agree
it would be ideal if we could just extend it.

File: llama_index/llms/konko.py
Comment (shivanimodi16) [reply llama_index/llms/konko.py:1]: Hi Logan,

Thank you for your feedback!

We did a double check on our end . While our API does share similarities with OpenAI, there are notable distinctions, such as supporting two API keys and our endpoint being specifically tailored towards `konko.completion`. Due to these differences, several utility functions required modifications. 

Using a separate implementation not only caters to our current requirements but also grants us the flexibility for potential future deviations from the OpenAI completion endpoint.

That said, we're always open to suggestions and improvements. If you have any recommendations or insights, we'd be more than happy to consider and incorporate them.

Comment (logan-markewich) [reply llama_index/llms/konko.py:1]: Thanks for checking!

**Comment by logan-markewich** [comment]
@mahaddad I can't push changes to this branch. Do you mind running the linting to fix CI?

```
pip install -r requirements.txt
black . 
ruff . --fix
```

**Comment by mahaddad** [comment]
Yeah, I can do that.  I'll ping you when it's done

**Comment by mahaddad** [comment]
Should be fixed, pending workflow approval.

**Comment by logan-markewich** [comment]
@mahaddad whoops, should have added mypy 😅 Just need to add return type annotations to a few functions in the tests you added

**Comment by mahaddad** [comment]
I see that it failed. Looking into it. It seemed to pass locally.

**Comment by mahaddad** [comment]
Okay should be good to go now 🙏 

Thank you!

**Comment by logan-markewich** [comment]
Awesome! Thanks for the quick iteration here. Merging now! (Apologies for the delay in getting this in!)

**Pull request merged**